Mathlib/CategoryTheory/Sites/LocallyInjective.lean

```diff
@@ -0,0 +1,196 @@
+the equality `x = y` must hold locally, i.e. after restriction
+by the maps of a covering sieve.
+
+-/
+
+universe w v' v u' u
+
+namespace CategoryTheory
+
+open Opposite Limits
+
+variable {C : Type u} [Category.{v} C]
+  {D : Type u'} [Category.{v'} D] [ConcreteCategory.{w} D]
+  (J : GrothendieckTopology C)
+
+attribute [local instance] ConcreteCategory.hasCoeToSort ConcreteCategory.instFunLike
+
+namespace Presheaf
+
+/-- If `F : Cᵒᵖ ⥤ D` is a presheaf with values in a concrete category, if `x` and `y` are
+elements in `F.obj X`, this is the sieve of `X.unop` consisting of morphisms `f`
+such that `F.map f.op x = F.map f.op y`. -/
+@[simps]
+def equalizerSieve {F : Cᵒᵖ ⥤ D} {X : Cᵒᵖ} (x y : F.obj X) : Sieve X.unop where
+  arrows _ f := F.map f.op x = F.map f.op y
+  downward_closed {X Y} f hf g := by
+    dsimp at hf ⊢
+    simp [hf]
+
+variable {F₁ F₂ F₃ : Cᵒᵖ ⥤ D} (φ : F₁ ⟶ F₂) (ψ : F₂ ⟶ F₃)
+
+/-- A morphism `φ : F₁ ⟶ F₂` of presheaves `Cᵒᵖ ⥤ D` (with `D` a concrete category)
+is locally injective for a Grothendieck topology `J` on `C` if
+whenever two sections of `F₁` are sent to the same section of `F₂`, then these two
+sections coincide locally. -/
+class IsLocallyInjective : Prop where
+  equalizerSieve_mem {X : Cᵒᵖ} (x y : F₁.obj X) (h : φ.app X x = φ.app X y) :
+    equalizerSieve x y ∈ J X.unop
+
+lemma equalizerSieve_mem [IsLocallyInjective J φ]
+    {X : Cᵒᵖ} (x y : F₁.obj X) (h : φ.app X x = φ.app X y) :
+    equalizerSieve x y ∈ J X.unop :=
+  IsLocallyInjective.equalizerSieve_mem x y h
+
+lemma isLocallyInjective_of_injective (hφ : ∀ (X : Cᵒᵖ), Function.Injective (φ.app X)) :
+    IsLocallyInjective J φ where
+  equalizerSieve_mem {X} x y h := by
+    convert J.top_mem X.unop
+    ext Y f
+    simp only [equalizerSieve_apply, op_unop, Sieve.top_apply, iff_true]
+    apply hφ
+    simp [h]
+
+instance [IsIso φ] : IsLocallyInjective J φ :=
+  isLocallyInjective_of_injective J φ (fun X => Function.Bijective.injective (by
+    rw [← isIso_iff_bijective]
+    change IsIso ((forget D).map (φ.app X))
+    infer_instance))
+
+instance isLocallyInjective_forget [IsLocallyInjective J φ] :
+    IsLocallyInjective J (whiskerRight φ (forget D)) where
+  equalizerSieve_mem x y h := equalizerSieve_mem J φ x y h
+
+instance isLocallyInjective_comp [IsLocallyInjective J φ] [IsLocallyInjective J ψ] :
+    IsLocallyInjective J (φ ≫ ψ) where
+  equalizerSieve_mem {X} x y h := by
+    let S := equalizerSieve (φ.app _ x) (φ.app _ y)
+    let T : ∀ ⦃Y : C⦄ ⦃f : Y ⟶ X.unop⦄ (_ : S f), Sieve Y := fun Y f _ =>
+      equalizerSieve (F₁.map f.op x) ((F₁.map f.op y))
+    refine J.superset_covering ?_
+      (J.transitive (equalizerSieve_mem J ψ (φ.app _ x) (φ.app _ y) (by simpa using h))
+      (Sieve.bind S.1 T) ?_)
+    · rintro Y f ⟨Z, a, g, hg, ha, rfl⟩
+      simpa using ha
+    · intro U f hf
+      exact J.superset_covering (Sieve.le_pullback_bind S.1 T _ hf)
+        (equalizerSieve_mem J φ (F₁.map f.op x) (F₁.map f.op y) (by simpa using hf))
```

Comment:
Maybe you can extract out the fact that `IsLocallyInjective J φ` iff `equalizerSieve (φ.app X x) (φ.app X y) ∈ J X.unop` implies `equalizerSieve x y ∈ J X.unop` from this proof?